design/rest-api-guideline.md

```diff
@@ -0,0 +1,388 @@
+some processing asynchronously without necessarily return a response body. In this case, developers can use `204` on `GET` responses to communicate to clients the result of the operation.
+
+#### POST
+
+HTTP `POST` should be used to create new resources. It is a non-safe and non-idempotent operation and clients should be aware of that.
+
+Successful `POST` requests should always result in a `201` status code and empty response. The response should also contain a reference
+to the resource that was created:
+
+```
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: http://mykeycloak.com/apis/users/{version}/1234
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `POST` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|201|
+|400|
+|401|
+|403|
+|404|
+|409|
+|500|
+
+#### PUT
+
+HTTP `PUT` should be used to replace a resource state.
+
+Successful `PUT` requests should always result in a `204` status code and empty response.
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `POST` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|204|
+|400|
+|401|
+|403|
+|404|
+|409|
+|500|
+
+#### DELETE
+
+HTTP `DELETE` should be used to remove a resource.
+
+Successful `DELETE` requests should always result in a `204` status code and empty response.
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `DELETE` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|204|
+|400|
+|401|
+|403|
+|404|
+|500|
+
+## Practices
+
+### Error Handling
+
+A good error handling mechanism helps to communicate to clients the reasons why a request could not be fulfilled and eventually
+recover from this errors.
+
+When returning errors, developers should use a common [HTTP Status Code](#http-status-codes) and include a body using JSON as the format with the information
+why a request failed.
+
+The error message should have the following format:
+
+|Claim|Description|Required|
+|---|---|---|
+|error|The error code|Yes|
+|error_description|A text providing more details for clients about the error and eventually how to recover from it|No|
+
+Here is an example of a error response:
+
+```json
+HTTP/1.1 400 Bad Request
+{
+    "error": "invalid_request",
+    "error_description": "The field 'email' is mandatory"
+}
+```
+
+#### Error Codes
+
+Error codes can be any string. They should be lowercase and use `_` when the code has multiple words. However, developers should 
+consider existing error codes before creating new ones.
+
+When processing error responses, developers should look the error codes being used by other APIs and see if they can be reused.
+
+### Filtering
```

Comment:
Is Filtering the same as search? I.e. if I want to find a user by username, or list users with a given role?

Reply:
Yes, it is.

Reply:
@stianst @pedroigor I would like to include the following content for filtering if possible:

### Filtering

Sometimes, we need more control over the content that will return from the server, so that only a subset of the data is available in the response. For example, a list user's groups or permissions. 

The Keycloak admin REST API allows to search for users based on `e-mail`, `first name`, `lastname` and other optional parameters. However, such implementation is limited to a particular context (e.g, collection of users).

Filtering can be implemented as a query parameter, like we already did for the current REST API, with some enhancements. To filter on a field, should be as simple as the following request:

```
/{realm}/apis/admin/users/v1?group=users1&realm=master&sort=desc
```

Multiple filters should result in an implicit `AND`, so in our example `/v1?group=users1&role=role1&realm=master&sort=desc` would provide results based on `groups`, `roles` and `realm`. Now let's say we want filter based on more than one groups:

```
/{realm}/apis/admin/users/v1?group=in:users1,users2&realm=master&sort=desc
```

A brief explanation about the parameters:

|Name|Type|Description|
|---|---|---|
|sort|string|Sort the results on our query based on: date, ascending order and descending order|
|in|string|Query for values inside a collection. For example: users belonging to group1 `and` group2.
|or|string|Query for values inside a collection. For example: users belonging to group1 `or` group2.

---
### Minor consideration

* It is possible to provide more generic queries like GitHub does in the future, by using something like `/commits?q=repo:octocat`. But for now, I'd vote to not over complicate things and stick with simple. At the moment we don't have anything that justifies generic queries in the Account REST API